source/common/network/connection_impl_base.cc

```diff
@@ -50,8 +50,13 @@ void ConnectionImplBase::initializeDelayedCloseTimer() {
 
 void ConnectionImplBase::raiseConnectionEvent(ConnectionEvent event) {
   for (ConnectionCallbacks* callback : callbacks_) {
-    // TODO(mattklein123): If we close while raising a connected event we should not raise further
-    // connected events.
+    // If a previous connected callback closed the connection, don't raise any further connected
+    // events. There was already recursion raising closed events.
+    if (event != ConnectionEvent::LocalClose && event != ConnectionEvent::RemoteClose &&
```

Comment:
Does it make sense to propagate multiple closed events?  I could imagine that being problematic as well but if we need it I think it's worth commenting why

Reply:
IMO it does, because people could be doing cleanup based on close. I will add more comments. At the very least, I suggest starting with this and we can always make it more restrictive later?